front: display exception summaries in the train list

clarani opened the pull request:
closes #12482

clarani committed:
front: enhance TimetableItemWithSummaries

Creates a new type SimulationSummary which distinguishes between invalid and
valid trains.

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: give specific rolling stock instead of rolling stocks list in formatBaseTimetableWithDetails

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: split formatBaseTimetableItemWithDetails into 2 functions

Split formatBaseTimetableItemWithDetails into 2 functions, one to handle
train schedules and one to handle paced trains.

This will be useful for the following commits.

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: handle the summary in formatTimetableItemWithDetails

Will be useful for the following commits.

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: store exception simulation in PacedTrainWithDetails

Each exception now has its own simulation in the pacedTrainWithDetails.

In the following commits, we will use the simulation of the exception when needed,
instead of the paced train one, which will allow us to display the simulation summary
of each exception in the train list.

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: display the simulation summary of paced train exceptions

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: display the correct simulation times for exception in output table

Get the path item times of the exception, instead of getting the ones of the paced
train.

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

SharglutDev commented on front/src/modules/timetableItem/components/Timetable/types.ts:
```diff
@@ -20,30 +20,32 @@ export type TrainCategoryFilter = 'all' | 'noCategory' | TrainMainCategory;
+  | { isValid: false; invalidReason: InvalidReason }
+  | {
+      isValid: true;
+      duration: Duration /** Travel time */;
```
nit : would be nicer like that so we can have the comment when hovering the property in other files
```suggestion
      /** Travel time */
      duration: Duration;
```

clarani replied:
Done

SharglutDev commented on front/src/modules/timesStops/hooks/useOutputTableData.ts:
```diff
@@ -97,7 +97,7 @@ const useOutputTableData = (
       };
     });
-  }, [selectedTrain, path, timetableItemWithDetails?.pathItemTimes]);
+  }, [selectedTrain, path, timetableItemWithDetails]);
```
This means all the computations in this `useMemo` will be done even when the simulation of the timetable item doesn't change (changing the name, label, category etc).

Could we do something for that ? Maybe `useOutputTableData` should accept only the prop : `timetableItemWithDetails?: Extract<TimetableItemWithDetails, { isValid: true }>,` (will change for no calcul issue) and then we can put back `timetableItemWithDetails.pathItemTimes` in the depths.

clarani replied:
I added a new commit to handle this problem, tell me if it's fine for you ;)

SharglutDev commented on front/src/modules/timetableItem/components/Timetable/types.ts:
```diff
@@ -20,30 +20,32 @@ export type TrainCategoryFilter = 'all' | 'noCategory' | TrainMainCategory;
-    final: SimulationSummaryResultSuccess['path_item_times_final'];
-  };
-};
+} & (({ isLoaded: true } & SimulationSummary) | { isLoaded: false; isValid: undefined });
```
Do we have to deal with `isLoaded` everywhere ? Won't it be possible to just use `isValid` ? (if `isValid` is not defined, it means the summaries haven't loaded yet)
```suggestion
} & (({ isValid: boolean } & SimulationSummary) | { isValid: undefined });
```

emersion replied:
Agreed. "isLoaded" is also a bit vague (what is loading? the train details, projection, or simulation?)

clarani replied:
I modified the type, is it better now ?

SharglutDev replied:
Works for me :)

anisometropie commented on front/src/modules/timetableItem/components/Timetable/useFilterTimetableItems.ts:
```diff
@@ -71,13 +71,13 @@ const useFilterTimetableItems = (
         // Apply scheduled points honored filter
         if (scheduledPointsHonoredFilter !== 'both') {
-          if (!timetableItem.isValid) {
+          if (!timetableItem.isLoaded || !timetableItem.isValid) {
```
`isValid` is undefined before it’s loaded, so `timetableItem.isValid` seems enough on its own

clarani replied:
Done

SharglutDev commented on front/src/modules/simulationResult/helpers/formatTimetableItemSummaries.ts:
```diff
@@ -1,33 +1,67 @@
-    if (!timetableItem) {
+  const items: TimetableItemWithDetails[] = [];
+
+  // train schedules
```
The code is exactly the same for both type if items. Can't we merge the logic ?

clarani replied:
Well, actually the code isn't exactly the same, because we're calling two different functions (`formatTrainScheduleWithDetails` and `formatPacedTrainWithDetails`) which require different arguments.
Even if I create a common function, I'd still need to check the type of the `timetableItem` 😕

Unless there's something I missed that you had in mind?

clarani replied:
Ok, I tried something, tell me if it's fine for you :)

SharglutDev replied:
Can't say I love this logic with the extra types and type guard but several days have passed since my first review, I can't really remember how it was before :) So we can keep it that way.

clarani replied:
Thanks :pray:

SharglutDev commented on front/src/modules/timetableItem/components/ManageTimetableItem/helpers/buildPacedTrainException.ts:
```diff
@@ -148,11 +148,11 @@ export function generatePacedTrainException(
-export function updatePacedTrainExceptionsList(
-  currentExceptions: PacedTrainException[],
-  newException: PacedTrainException,
+export function updatePacedTrainExceptionsList<T extends PacedTrainException>(
```
nit : as we saw in previous PR, typescript is perfectly fine passing an extended object to a function.
We could have kept `PacedTrainException` as type and it wouldn't have generated any error.

We should consider using this syntax for other part of the app for harmonization it you chose to keep it :)

anisometropie replied:
it’s not just a matter of « the function can accept more fields than the signature expects »
here T is also used as returned value, so we keep the info about the extra fields in type the function returns.

it’s also not just a matter of what TS can do, it’s a matter of what we want to express and mean with our code.

SharglutDev replied:
You're totally right and I like this accuracy. Nevertheless, there are other functions where we could be more precise about what we want to put inside the function and return like we do here.

clarani replied:
Hum so I don't have to change anything here ? :)

SharglutDev replied:
No it's fine like that :)

SharglutDev commented on front/src/modules/timetableItem/helpers/formatTimetableItemWithDetails.ts:
```diff
@@ -16,36 +18,28 @@ import { Duration } from 'utils/duration';
+  summary: Extract<SimulationSummaryResult, { status: 'success' }>
+) => {
   let notHonoredReason: Extract<TimetableItemWithDetails, { isValid: true }>['notHonoredReason'];
   if (summary.status === 'success') {
```
Not need anymore
```suggestion

```

clarani replied:
Done

anisometropie commented on front/src/modules/timetableItem/helpers/formatTimetableItemWithDetails.ts:
```diff
@@ -0,0 +1,142 @@
+
+  return {
+    ...pacedTrainProps,
+    name: pacedTrain.train_name,
+    startTime: new Date(pacedTrain.start_time),
+    stopsCount:
+      (pacedTrain.schedule?.filter((step) => step.stop_for && Duration.parse(step.stop_for).ms > 0)
+        .length ?? 0) + 1, // +1 to take the final stop (destination) into account
+    speedLimitTag: pacedTrain.speed_limit_tag ?? null,
+    labels: pacedTrain.labels ?? [],
+    rollingStock,
+    paced: {
+      timeWindow: Duration.parse(paced.time_window),
+      interval: Duration.parse(paced.interval),
+    },
```
use your destructured values from line 99 and remove eslint comment

clarani replied:
I added a new function, so I can't remove the eslint warning now :/

anisometropie replied:
if you just want to ignore them then make it clear.

we don’t want to create confusion reading the code in 6 months (which one is unused ? where are those variables going ? was some code removed  ?  is it just forgotten ?) and create more work to figure it out.

`const { train_name: _, start_time: __, speed_limit_tag: ___, ...trainScheduleProps } = trainSchedule;
`

clarani replied:
Done

anisometropie commented on front/src/modules/timetableItem/helpers/formatTimetableItemWithDetails.ts:
```diff
@@ -0,0 +1,142 @@
+  rollingStock?: LightRollingStockWithLiveries,
+  summary?: SimulationSummaryResult
+): TimetableItemWithDetails => {
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const { train_name, start_time, speed_limit_tag, ...trainScheduleProps } = timetableItem;
+
+  return {
+    ...trainScheduleProps,
+    name: timetableItem.train_name,
+    startTime: new Date(timetableItem.start_time),
+    stopsCount:
+      (timetableItem.schedule?.filter(
+        (step) => step.stop_for && Duration.parse(step.stop_for).ms > 0
+      ).length ?? 0) + 1, // +1 to take the final stop (destination) into account
+    speedLimitTag: timetableItem.speed_limit_tag ?? null,
+    labels: timetableItem.labels ?? [],
+    rollingStock,
+    ...formatSummary(timetableItem, summary),
+  };
+};
```
use train_name, start_time, speed_limit_tag from l.75

clarani replied:
I added a new function, so I can't remove the eslint warning now :/

anisometropie replied:
same as above

clarani replied:
Done

anisometropie commented on front/src/modules/simulationResult/helpers/formatTimetableItemSummaries.ts:
```diff
@@ -1,68 +1,67 @@
+      throw new Error('Missing timetable item');
+    }
+    if (!isPacedTrainResponseWithPacedTrainId(pacedTrain)) {
+      throw new Error('Timetable item should be a paced train, but it is a train schedule');
```
```suggestion
      throw new Error('Timetable item should be a paced train');
```
we just know its id doesn’t start with 'paced_' from the test we do. it could be any kind of error. better not assume what it is and avoid misleading when debugging

clarani replied:
Done

SharglutDev commented on front/src/modules/timetableItem/components/Timetable/types.ts:
```diff
@@ -94,6 +94,16 @@ export type ExceptionChangeGroups = Omit<
   'key' | 'occurrence_index' | 'disabled'
 >;
 
+export type OccurrenceSummary =
```
Shouldn't the occurrences also have an `notHonoredReason` ?

clarani replied:
Ah yes, you're right !

anisometropie commented on front/src/modules/timetableItem/components/Timetable/PacedTrain/hooks/useOccurrences.ts:
```diff
@@ -105,32 +106,30 @@ const useOccurrences = (
-          : {
-              isValid: false,
-            }),
+        exceptionChangeGroups: omit(exception, [
+          'key',
+          'disabled',
+          'occurrence_index',
+          'simulation',
+        ]),
```
what’s the deal with simulation exception change group?

clarani replied:
This exceptionChangeGroups is used to display the list of the exceptionChangeGroups of the occurrence. To compute it, we take all the keys of the exception, but some of them are not change groups (such as simulation), so we need to remove them manually

anisometropie commented on front/src/modules/timesStops/hooks/useOutputTableData.ts:
```diff
@@ -36,12 +41,27 @@ const useOutputTableData = (
+        ? timetableItemWithDetails.pathItemTimes
+        : undefined;
+    } else {
+      if (isTrainScheduleId(selectedTrain.id)) throw new Error();
```
add info into the error about what you expect to get,

clarani replied:
Done

SharglutDev commented on front/src/modules/timetableItem/components/Timetable/PacedTrain/hooks/useOccurrences.ts:
```diff
@@ -15,7 +15,26 @@ import {
+  SimulationSummary,
+} from '../../types';
+
+const extractSummaryData = (summary: SimulationSummary, startTime: Date): OccurrenceSummary =>
```
To keep your logic on the other function, here we should use `T extends SimulationSummary`

clarani replied:
I removed this function ;)

anisometropie reviewed (changes requested):
great work making Summary types more robust and better represent valid state. also handling train schedule and paced train separately is really an improvement

SharglutDev commented on front/src/modules/timetableItem/helpers/formatTimetableItemWithDetails.ts:
```diff
@@ -21,14 +77,14 @@ export const formatTrainScheduleWithDetails = (
   };
 };
 
 export const formatPacedTrainWithDetails = (
```
After all, the "base" of these functions is still the same, we could have kept the "base formatting function" to avoid repeating ourselves.

clarani replied:
I added a function `extractBaseTimetableItemProps` :+1:

SharglutDev commented on front/src/modules/timesStops/hooks/useOutputTableData.ts:
```diff
@@ -36,12 +41,27 @@ const useOutputTableData = (
+        timetableItemWithDetails.exceptions,
+        selectedTrain.id
+      );
+      const summary = exception ? exception.simulation : timetableItemWithDetails;
```
Feels very weird. Maybe it would have been cleaner and more harmonized to have the summary of the paced train also in a `simulation` property.

clarani replied:
Yes, done :white_check_mark:

SharglutDev reviewed (commented):
Great job, not tested yet.

emersion commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -130,9 +133,13 @@ const useScenarioData = (scenario: ScenarioResponse, infra: InfraWithState) => {
-        simulatedTrain ?? formatBaseTimetableItemWithDetails(timetableItem, rollingStocks ?? [])
+      if (simulatedTrain) return simulatedTrain;
+      const rollingStock = (rollingStocks ?? []).find(
+        (rs) => rs.name === timetableItem.rolling_stock_name
```
Note, we should be able to use `rollingStocksMap` instead of `rollingStocks` now. (Feel free to resolve if you want to keep this for later.)

clarani replied:
Done for useScenarioData, but not in `formatTimetableItemSummaries` because this will require some change in STDCM :+1:

Synar reviewed (approved):
e2e LGTM ^^

clarani committed:
front: use SimulationSummary instead of TimetableItemWithDetails in useOutputTableData

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

SharglutDev commented on front/src/modules/timetableItem/helpers/formatTimetableItemWithDetails.ts:
```diff
@@ -0,0 +1,50 @@
+  rollingStock?: LightRollingStockWithLiveries
+): TimetableItemWithDetails => {
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const { train_name, start_time, speed_limit_tag, ...trainScheduleProps } = trainSchedule;
```
Great that you remove these unnecessary fields 🎉 

Since a `TimetableItemWithDetails` is supposed to omit `rolling_stock_name` and `timetable_id`, would you mind removing them as well ? (same in `formatPacedTrainWithDetails`)

clarani replied:
I removed `rolling_stock_name`, but `TrainScheduleWithTrainId` and `PacedTrainWithPacedTrainId` don't have `timetable_id`

SharglutDev replied:
Actually the `trainSchedule` param here (same for paced train) comes from `TrainScheduleResponse` which contains `timetable_id`. Since Typescript doesn't complain with extra properties in a type, this went silent when doing `setTimetableItems` in `useScenarioData`.

<img width="276" height="321" alt="Capture d’écran 2025-07-29 à 08 25 29" src="https://github.com/user-attachments/assets/d5a13579-85ef-4a8e-aa7b-46f93e0a7c96" />

clarani replied:
Can we leave it for another PR since it's not related to mine ? :angel:

SharglutDev replied:
Haha sure 🙂

SharglutDev commented on front/src/modules/simulationResult/helpers/formatTimetableItemSummaries.ts:
```diff
@@ -1,33 +1,78 @@
-/**
- * Format the timetable items with their simulation summaries
- */
+type Inputs =
```
`Inputs` is pretty confusing to me 🤔 What kind of inputs are we talking about ? `summary` is rather an output no ?

clarani replied:
Is it clearer now ? I think the name of this type is not very important because it's an internal type

SharglutDev commented on front/src/modules/simulationResult/helpers/formatTimetableItemSummaries.ts:
```diff
@@ -1,33 +1,78 @@
+  }
+  const rollingStock = rollingStocks.find((rs) => rs.name === timetableItem.rolling_stock_name);
+
+  if (isPacedTrainResponseWithPacedTrainId(timetableItem) && isPacedTrainInputs(inputs))
```
nit : in both other `if` in the function you use it with the `{}` syntax. Could we do the same here for harmonization ? :)

clarani replied:
Done

SharglutDev commented on front/src/modules/simulationResult/helpers/formatTimetableItemSummaries.ts:
```diff
@@ -1,33 +1,78 @@
+    return formatPacedTrainWithDetails(timetableItem, rollingStock, inputs.summary);
+
+  if (isPacedTrainInputs(inputs) || isPacedTrainResponseWithPacedTrainId(timetableItem)) {
+    throw new Error('Mismatch bewteen timetableItemId and timetableItem');
```
```suggestion
    throw new Error('Mismatch between timetableItemId and timetableItem');
```

clarani replied:
Done

SharglutDev commented on front/src/modules/timetableItem/helpers/formatTimetableItemWithDetails.ts:
```diff
@@ -16,26 +18,21 @@ import { Duration } from 'utils/duration';
+  return summary.status === 'pathfinding_not_found' || summary.status === 'pathfinding_input_error'
+    ? summary.error_type
+    : summary.status;
+};
```
Both times where we call this function, we check for summary.status === 'success', is it really necessary to add another check here ?

We could rather do : 
```suggestion
const extractInvalidReason = (summary: Exclude<SimulationSummaryResult, { status: 'success' }>) =>
  summary.status === 'pathfinding_not_found' || summary.status === 'pathfinding_input_error'
    ? summary.error_type
    : summary.status;
```

clarani replied:
You're right, thanks :)

SharglutDev commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -132,7 +135,9 @@ const useScenarioData = (scenario: ScenarioResponse, infra: InfraWithState) => {
       if (simulatedTrain) return simulatedTrain;
       const rollingStock = rollingStocksByName.get(timetableItem.rolling_stock_name);
-      return formatBaseTimetableItemWithDetails(timetableItem, rollingStock);
+      return isPacedTrainResponseWithPacedTrainId(timetableItem)
+        ? formatPacedTrainWithDetails(timetableItem, rollingStock)
+        : formatTrainScheduleWithDetails(timetableItem, rollingStock);
```
Do we still have to format here since we handle a `simulatedTrain` which should have already been formatted in `useLazySimulateTrains` ?

clarani replied:
This is used to populate timetableItemsWithDetails in case the simulation of the item is not loaded yet, so it would not have been formatted in `useLazySimulationTrains`. We need to keep it I think :+1:

SharglutDev commented on front/src/applications/operationalStudies/hooks/useSimulationResults.ts:
```diff
@@ -128,7 +128,7 @@ const useSimulationResults = (infraId: number): SimulationResults | undefined =>
     !rollingStock
   ) {
-    return undefined;
+    return { isValid: false, train: train ?? undefined, rollingStock: rollingStock ?? undefined };
```
You can simplify this
```suggestion
    return { isValid: false, train, rollingStock };
```

clarani replied:
I dropped the commit ;)

SharglutDev commented on front/src/applications/operationalStudies/hooks/useSimulationResults.ts:
Are these changes for the non calcul in the output table ? If yes, I think it will be handled differently.

clarani replied:
I'm going to drop this commit :+1:

SharglutDev reviewed (commented):
Still haven't had the time to test 🫣

anisometropie reviewed (approved):
tested

SharglutDev reviewed (commented):
Tested the PR. Everything works as expected except for a train with a trackoffset where the output table is on infinite load.

SharglutDev commented:
One last issue when testing, it's not possible to select an added exception in the timetable.

clarani committed:
front: make added exceptions selectable

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

SharglutDev commented:
Another small issue :) 

https://github.com/user-attachments/assets/6293391d-e342-4ef0-aa59-859fd85ca422

clarani committed:
front: fix occurrence start time

The occurrence start_time was overwritten by the function extractOccurrenceDetailsFromPacedTrain.
Now the correct start_time is displayed.

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

SharglutDev reviewed (approved):
Lgtm and tested ✅ Great job !
